src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/ExactMethodInstantiationsNode.cs

```diff
@@ -0,0 +1,169 @@
+        {
+            if (!IsMethodEligibleForTracking(method))
+                return null;
+
+            DependencyList dependencies = new DependencyList();
+
+            dependencies.Add(new DependencyListEntry(factory.NecessaryTypeSymbol(method.OwningType), "Exact method instantiation signature"));
+
+            foreach (var arg in method.Instantiation)
+                dependencies.Add(new DependencyListEntry(factory.NecessaryTypeSymbol(arg), "Exact method instantiation signature"));
+
+            return dependencies;
+        }
+
+        public void AddExactMethodInstantiationEntry(NodeFactory factory, MethodDesc method)
+        {
+            // Check if we already saw this method
+            if (!_exactMethodInstantiationsList.Add(method))
+                return;
+
+            if (!IsMethodEligibleForTracking(method))
+                return;
+
+            bool getUnboxingStub = (method.OwningType.IsValueType || method.OwningType.IsEnum) && !method.Signature.IsStatic;
+            IMethodNode methodEntryPointNode = factory.MethodEntrypoint(method, getUnboxingStub);
+
+            Vertex methodSignature = GetNativeLayoutInfoSignatureForMethod(factory, method);
+            Vertex methodPointer = _nativeWriter.GetUnsignedConstant(_externalReferences.GetIndex(methodEntryPointNode));
+
+            // Make the generic method entry vertex
+
+            Vertex entry = _nativeWriter.GetTuple(methodSignature, methodPointer);
+
+            // Add to the hash table, hashed by the containing type's hashcode
+            uint hashCode = (uint)method.OwningType.GetHashCode();
+            _hashtable.Append(hashCode, _nativeSection.Place(entry));
+        }
+
+        private bool IsMethodEligibleForTracking(MethodDesc method)
+        {
+            if (method.IsAbstract)
+                return false;
+
+            if (!method.HasInstantiation)
+                return false;
+
+            // Nothing to add for interface methods because they have no implementations of their own...
+            if (method.OwningType.IsInterface)
+                return false;
+
+            if (method.IsRuntimeDeterminedExactMethod)
+                return false;
+
+            // This hashtable is only for method instantiations that don't use generic dictionaries,
+            // so check if the given method is shared before proceeding
+            if (method.IsSharedByGenericInstantiations || method.GetCanonMethodTarget(CanonicalFormKind.Specific) != method)
+                return false;
+
+            return true;
+        }
+
+        private Vertex GetNativeLayoutInfoSignatureForMethod(NodeFactory factory, MethodDesc method)
+        {
+            // Get native layout vertices for the declaring type
+
+            ISymbolNode declaringTypeNode = factory.NecessaryTypeSymbol(method.OwningType);
+            Debug.Assert(declaringTypeNode != null);
```

Comment:
Nit: Null is never a valid thing to return from `NecessaryTypeSymbol`. Asserting would only make sense if it's a valid return value under certain circumstances, but we don't expect that value here.